hypershift-operator/controllers/controlplaneautoscaler/controller.go

```diff
@@ -0,0 +1,349 @@
+		size   string
+		memory resource.Quantity
+	}
+	sizesToOrder := make([]sizeWithMemory, 0, len(s.sizes))
+	for size, resources := range s.sizes {
+		sizesToOrder = append(sizesToOrder, sizeWithMemory{
+			size:   size,
+			memory: resources.Memory,
+		})
+	}
+	slices.SortFunc(sizesToOrder, func(a, b sizeWithMemory) int {
+		return a.memory.Cmp(b.memory)
+	})
+	sortedSizes := make([]string, len(sizesToOrder))
+	for i := range sizesToOrder {
+		sortedSizes[i] = sizesToOrder[i].size
+	}
+	return sortedSizes
+}
+
+func SetupWithManager(mgr ctrl.Manager) error {
+	controller := &ControlPlaneAutoscalerController{
+		Client: mgr.GetClient(),
+	}
+	if err := ctrl.NewControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&hyperv1.HostedCluster{}).
+		Watches(&vpaautoscalingv1.VerticalPodAutoscaler{}, handler.EnqueueRequestsFromMapFunc(hostedClusterForVPA)).
+		Complete(controller); err != nil {
+		return err
+	}
+	return nil
+}
+
+func hostedClusterForVPA(ctx context.Context, object client.Object) []reconcile.Request {
+	vpa, isVPA := object.(*vpaautoscalingv1.VerticalPodAutoscaler)
+	if !isVPA {
+		return nil
+	}
+	ns := vpa.Annotations[hcNamespaceAnnotation]
+	name := vpa.Annotations[hcNameAnnotation]
+	if ns == "" || name == "" {
+		return nil
+	}
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{
+				Namespace: ns,
+				Name:      name,
+			},
+		},
+	}
+}
+
+func (r *ControlPlaneAutoscalerController) UpdateSizeCache(ctx context.Context) error {
+	config := &schedulingv1alpha1.ClusterSizingConfiguration{}
+	config.Name = "cluster"
+	if err := r.Get(ctx, client.ObjectKeyFromObject(config), config); err != nil {
+		return fmt.Errorf("failed to fetch cluster sizing configuration: %w", err)
+	}
+
+	listMachineSets := func() (*machinev1beta1.MachineSetList, error) {
+		result := &machinev1beta1.MachineSetList{}
+		err := r.List(ctx, result, client.InNamespace("openshift-machine-api"))
+		return result, err
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	if err := r.sizeCache.update(config, listMachineSets, log); err != nil {
+		return fmt.Errorf("failed to update machine size cache: %w", err)
+	}
+	return nil
+}
+
+func (r *ControlPlaneAutoscalerController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	if err := r.UpdateSizeCache(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	hc := &hyperv1.HostedCluster{}
+	if err := r.Get(ctx, request.NamespacedName, hc); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if !hc.DeletionTimestamp.IsZero() {
+		// No need to process deleting HostedClusters
+		return ctrl.Result{}, nil
+	}
+
+	cpNamespace := manifests.HostedControlPlaneNamespace(hc.Namespace, hc.Name)
+	vpa := controlplaneautoscalermanifests.KubeAPIServerVerticalPodAutoscaler(cpNamespace)
+
+	if hc.Annotations[hyperv1.ControlPlaneAutoscalingAnnotation] != "true" || hc.Annotations[hyperv1.TopologyAnnotation] != hyperv1.DedicatedRequestServingComponentsTopology {
```

Comment:
should Reconcile just return earlier if TopologyAnnotation != dedicated?

Reply:
yes, but also if we're not automatically scaling the control plane, we need both conditions